libraries/freertos_plus/standard/https/src/iot_https_client.c

```diff
@@ -0,0 +1,2036 @@
+                        IOT_HTTPS_MAX_ALPN_PROTOCOLS_LENGTH );
+                    IOT_SET_AND_GOTO_CLEANUP( IOT_HTTPS_INVALID_PARAMETER );
+                }
+                memcpy( pAlpnProtos, pConnInfo->pAlpnProtocols, pConnInfo->alpnProtocolsLen );
+                pAlpnProtos[pConnInfo->alpnProtocolsLen] = '\0';
+                networkCredentials.pAlpnProtos = pAlpnProtos; /* This requires a NULL termination. It is inconsistent with other members in the struct. */
+            }
+            else
+            {
+                networkCredentials.pAlpnProtos = NULL;
+            }
+
+            /* If any of these are NULL a network error will result depending on the connection. */
+            networkCredentials.pRootCa = pConnInfo->pCaCert;
+            networkCredentials.rootCaSize = pConnInfo->caCertLen;
+            networkCredentials.pClientCert = pConnInfo->pClientCert;
+            networkCredentials.clientCertSize = pConnInfo->clientCertLen;
+            networkCredentials.pPrivateKey = pConnInfo->pPrivateKey;
+            networkCredentials.privateKeySize = pConnInfo->privateKeyLen;
+        }
+    }
+
+    /* If this is a TLS connection connect with networkCredentials. Otherwise pass NULL. */
+    if( ( pConnInfo->flags & IOT_HTTPS_IS_NON_TLS_FLAG ) == 0 )
+    {
+        /* create() will connect to the server specified. */
+        networkStatus = _httpsConnection->pNetworkInterface->create( &networkServerInfo,
+            &networkCredentials,
+            &( _httpsConnection->pNetworkConnection ) );
+    }
+    else
+    {
+        networkStatus = _httpsConnection->pNetworkInterface->create( &networkServerInfo,
+            NULL,
+            &( _httpsConnection->pNetworkConnection ) );
+    }
+
+    /* Check to see if the network connection succeeded. If we did not succeed then the connHandle
+       will be NULL and we return an error. */
+    if( networkStatus != IOT_NETWORK_SUCCESS )
+    {
+        IotLogError( "Failed to connect to the server at %.*s on port %d with error: %d",
+            pConnInfo->addressLen,
+            pConnInfo->pAddress,
+            pConnInfo->port,
+            networkStatus );
+        IOT_SET_AND_GOTO_CLEANUP(IOT_HTTPS_CONNECTION_ERROR);
+    }
+    else
+    {
+        /* The connection succeeded so this a connected context. */
+        _httpsConnection->isConnected = true;
+    }
+
+    /* The receive callback tells the task context handling the request/response that the network
+       is ready to read from. */
+    networkStatus = _httpsConnection->pNetworkInterface->setReceiveCallback( _httpsConnection->pNetworkConnection,
+        _networkReceiveCallback,
+        _httpsConnection );
+    if( networkStatus != IOT_NETWORK_SUCCESS )
+    {
+        IotLogError( "Failed to connect to set the HTTPS receive callback. " );
+        IOT_SET_AND_GOTO_CLEANUP(IOT_HTTPS_INTERNAL_ERROR);
+    }
+
+    /* Connection was successful, so create semaphores. */
```

Comment:
General Comment/ question : We are using multiple semaphores to make the application 'thread safe' 
1. Do we have a real need to share a network connection? If answer is no, we should not allow connection to be shared by multiple threads, it will make code simpler. Thread can always open a new connection and then use the library.
2.  Event flags is probably better option (although not yet available for CSDK), if we really want multiple threads to share the same connection.
3. We need to have integration tests to make sure this code works in multiple threads scenario as intended.

Reply:
1. This is a good point. There is not need to be thread safe, but that seems to be the direction that is preferred. I am going to keep the thread safetey and make waiting on the semaphore a bit simpler.
2. Yes that has been realized, but it is not currently available and implementing that is another feature.
3. That is being worked on. branch: feat/http_test